runtime/mount/python_lib/lplots/widgets/tabular_datasource.py

```diff
@@ -0,0 +1,121 @@
+from dataclasses import dataclass
+from typing import Any, Literal, NotRequired, TypedDict
+
+import latch_data_validation
+import latch_data_validation.data_validation
+import pandas as pd
+from latch.ldata.path import LPath
+from latch.registry.table import Table
+
+from .. import _inject
+from ..reactive import Signal
+from . import _emit, _state
+from .shared import FormInputAppearance
+
+DataSourceType = Literal["ldata", "dataframe", "registry"]
+
+
+class LDataDataSource(TypedDict):
+    node_id: str
+    type: Literal["ldata"]
+
+
+class DataFrameDataSource(TypedDict):
+    df_id: str
```

Comment:
probably don't want df_id cz i can do `kek  = pd.read_csv(...)` df_id suggest it takes in `df_{num}`

Reply:
Huh?